modules/AOR_Reports/aor_utils.php

```diff
@@ -570,3 +585,19 @@ function convertToDateTime($value)
 
     return $dateTime;
 }
+
+function setMinTime(DateTime $dt){
+    if (version_compare(PHP_VERSION, '7.1.0') >= 0) {
+        $dt->setTime(0, 0, 0, 0);
+    } else {
+        $dt->setTime(0, 0, 0);
+    }
+}
+
+function setMaxTime(DateTime $dt){
+    if (version_compare(PHP_VERSION, '7.1.0') >= 0) {
+        $dt->setTime(23, 59, 59, 999999);
+    } else {
+        $dt->setTime(23, 59, 59);
+    }
+}
```

Comment:
Seeing as this is used by just reports, do we really need to add the overhead of running version compares to include micro-seconds?

```suggestion
/**
 * @param DateTime $dt
 */
function setMaxTime(DateTime $dt)
{
    $dt->setTime(23, 59, 59);
}
```

Reply:
My thought is from the DB perspective, as 59 vs 59.999999 are different.

I modeled it similar to LocalDateTime which has min/max constants but in PHP microseconds have only been supported after 7.1. If minimum ver was set then it could be simplified.

Reply:
True, however I believe the only places we call ```getPeriodEndDate``` and ```getPeriodDate``` are here: 

https://github.com/salesagility/SuiteCRM/blob/hotfix-7.10.x/modules/AOR_Reports/AOR_Report.php#L1896
https://github.com/salesagility/SuiteCRM/blob/hotfix-7.10.x/modules/AOR_Reports/AOR_Report.php#L1740
https://github.com/salesagility/SuiteCRM/blob/hotfix-7.10.x/modules/AOR_Reports/AOR_Report.php#L1897

Which all format down to the second. Also it appears as though the only method to call ```calculateQuarters()```

https://github.com/salesagility/SuiteCRM/blob/hotfix-7.10.x/modules/AOR_Reports/aor_utils.php#L238

Is returning with:

```php
    $datetime_period = $datetime_period->setTime(0, 0, 0);

    return $timedate->tzGMT($datetime_period);
```

So unless I'm mistaken there isn't actually anywhere that will run against the DB using ```999999``` even if it's present.

Reply:
We can not for sure know what data users are writing directly to the DB or the column format they are using, this is just a belts and braces approach to make sure that queries against the DB are inclusive of all rows even if people are using internal methods.

Sub-second accuracy especially when you want to make sure audit logs are in order is important, this just covers the bases.

Ive got a custom module that is externally populated and tracks external events, to maintain order im using subsecond timestamps, but then we report on it so need to make sure that its correct.

Of course this would not be a problem if the SQL query was "date >= XXXX and < XXXX (rounded up to next second)` but when its using `<=` or `BETWEEN` the problem can become apparent.

Not that we use PG for SuiteCRM (though a few have asked for that compatibility to be added in the past), sub second accuracy is fairly common place with datetime field.

To that end its futureproofing IMO.

Reply:
This post pointed to the prior issue of pushing the second forward due to use of `BETWEEN` so I corrected it with microsecond precision. We were having reports including things that were exactly on `00:00:00.000000`.

https://github.com/salesagility/SuiteCRM/pull/8764#issuecomment-818815096

Reply:
Yeah I see your point, in that case I would suggest the following:

```php
/**
 * @param DateTime $dt
 * @noinspection DateTimeSetTimeUsageInspection
 */
function setMaxTime(DateTime $dt)
{
    if (PHP_VERSION_ID >= 70100) {
        $dt->setTime(23, 59, 59, 999999);
    } else {
        $dt->setTime(23, 59, 59);
    }
}
```

Just to remove the function call and set a noinspection. It may also be worth adding even a basic test for this so we can check nothing crazy happens across PHP versions, maybe something like:

```php
use SuiteCRM\Test\SuitePHPUnitFrameworkTestCase;

require_once __DIR__ . '/../../../../../modules/AOR_Reports/aor_utils.php';

/**
 * Class aor_utilsTest
 */
class aor_utilsTest extends SuitePHPUnitFrameworkTestCase
{
    public function testsetMaxTime()
    {
        $dateTime = DateTime::createFromFormat(
            'Y-m-d H:i:s',
            '1970-01-01 00:00:00'
        );
        setMaxTime($dateTime);
        $timeStamp = $dateTime->getTimestamp();

        self::assertEquals(86399, $timeStamp);
    }
}
```

Inside ```tests/unit/phpunit/modules/Aor_reports/aor_utilsTest.php```

Reply:
Ive modified and added UT